File: objectModel/CSharp/Microsoft.CommonDataModel.ObjectModel/Persistence/CdmFolder/Projections/ProjectionPersistence.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

namespace Microsoft.CommonDataModel.ObjectModel.Persistence.CdmFolder
{
    using Microsoft.CommonDataModel.ObjectModel.Cdm;
    using Microsoft.CommonDataModel.ObjectModel.Enums;
    using Microsoft.CommonDataModel.ObjectModel.Persistence.CdmFolder.Types;
    using Microsoft.CommonDataModel.ObjectModel.Utilities;
    using Microsoft.CommonDataModel.ObjectModel.Utilities.Logging;
    using Newtonsoft.Json.Linq;
    using System.Collections.Generic;

    /// <summary>
    /// Projection persistence
    /// </summary>
    public class ProjectionPersistence
    {
        public static CdmProjection FromData(CdmCorpusContext ctx, JToken obj)
        {
            if (obj == null)
            {
                return null;
            }

            CdmProjection projection = ctx.Corpus.MakeObject<CdmProjection>(CdmObjectType.ProjectionDef);

            CdmEntityReference source = EntityReferencePersistence.FromData(ctx, obj["source"]);

            if (obj["explanation"] != null)
            {
                projection.Explanation = (string)obj["explanation"];
            }

            projection.Condition = obj["condition"]?.ToString();
            projection.RunSequentially = (bool?)obj["runSequentially"];

            if (obj["operations"] != null)
            {
                List<JObject> operationJsons = obj["operations"]?.ToObject<List<JObject>>();
                foreach (JObject operationJson in operationJsons)
                {
                    string type = (string)operationJson["$type"];
                    switch (type)
                    {
                        case "addCountAttribute":
                            CdmOperationAddCountAttribute addCountAttributeOp = OperationAddCountAttributePersistence.FromData(ctx, operationJson);
                            projection.Operations.Add(addCountAttributeOp);
                            break;
                        case "addSupportingAttribute":
                            CdmOperationAddSupportingAttribute addSupportingAttributeOp = OperationAddSupportingAttributePersistence.FromData(ctx, operationJson);
                            projection.Operations.Add(addSupportingAttributeOp);
                            break;
                        case "addTypeAttribute":
                            CdmOperationAddTypeAttribute addTypeAttributeOp = OperationAddTypeAttributePersistence.FromData(ctx, operationJson);
                            projection.Operations.Add(addTypeAttributeOp);
                            break;
                        case "excludeAttributes":
                            CdmOperationExcludeAttributes excludeAttributesOp = OperationExcludeAttributesPersistence.FromData(ctx, operationJson);
                            projection.Operations.Add(excludeAttributesOp);
                            break;
                        case "arrayExpansion":
                            CdmOperationArrayExpansion arrayExpansionOp = OperationArrayExpansionPersistence.FromData(ctx, operationJson);
                            projection.Operations.Add(arrayExpansionOp);
                            break;
                        case "combineAttributes":
                            CdmOperationCombineAttributes combineAttributesOp = OperationCombineAttributesPersistence.FromData(ctx, operationJson);
                            projection.Operations.Add(combineAttributesOp);
                            break;
                        case "renameAttributes":
                            CdmOperationRenameAttributes renameAttributesOp = OperationRenameAttributesPersistence.FromData(ctx, operationJson);
                            projection.Operations.Add(renameAttributesOp);
                            break;
                        case "replaceAsForeignKey":
                            CdmOperationReplaceAsForeignKey replaceAsForeignKeyOp = OperationReplaceAsForeignKeyPersistence.FromData(ctx, operationJson);
                            projection.Operations.Add(replaceAsForeignKeyOp);
                            break;
                        case "includeAttributes":
                            CdmOperationIncludeAttributes includeAttributesOp = OperationIncludeAttributesPersistence.FromData(ctx, operationJson);
                            projection.Operations.Add(includeAttributesOp);
                            break;
                        case "addAttributeGroup":
                            CdmOperationAddAttributeGroup addAttributeGroupOp = OperationAddAttributeGroupPersistence.FromData(ctx, operationJson);
                            projection.Operations.Add(addAttributeGroupOp);
                            break;
                        default:
                            Logger.Error(nameof(ProjectionPersistence), ctx, $"Invalid operation type '{type}'.", nameof(FromData));
                            break;
                    }
                }
            }

            projection.Source = source;

            return projection;
        }

        public static Projection ToData(CdmProjection instance, ResolveOptions resOpt, CopyOptions options)
        {
            if (instance == null)
            {
                return null;
            }

            dynamic source = null;
            if (instance.Source != null && instance.Source.GetType() == typeof(string))
            {
                source = instance.Source;
            }
            else if (instance.Source != null && !string.IsNullOrWhiteSpace(instance.Source.NamedReference) && instance.Source.ExplicitReference == null)
            {
                source = instance.Source.NamedReference;
            }
            else if (instance.Source != null && instance.Source.GetType() == typeof(CdmEntityReference))
            {
                source = EntityReferencePersistence.ToData(instance.Source, resOpt, options);
            }

            List<OperationBase> operations = null;
            if (instance.Operations != null && instance.Operations.Count > 0)
            {
                operations = new List<OperationBase>();
                foreach (CdmOperationBase operation in instance.Operations)
                {
                    switch (operation.ObjectType)
                    {
                        case CdmObjectType.OperationAddCountAttributeDef:
                            OperationAddCountAttribute addCountAttributeOp = OperationAddCountAttributePersistence.ToData(operation as CdmOperationAddCountAttribute, resOpt, options);
                            operations.Add(addCountAttributeOp);
                            break;
                        case CdmObjectType.OperationAddSupportingAttributeDef:
                            OperationAddSupportingAttribute addSupportingAttributeOp = OperationAddSupportingAttributePersistence.ToData(operation as CdmOperationAddSupportingAttribute, resOpt, options);
                            operations.Add(addSupportingAttributeOp);
                            break;
                        case CdmObjectType.OperationAddTypeAttributeDef:
                            OperationAddTypeAttribute addTypeAttributeOp = OperationAddTypeAttributePersistence.ToData(operation as CdmOperationAddTypeAttribute, resOpt, options);
                            operations.Add(addTypeAttributeOp);
                            break;
                        case CdmObjectType.OperationExcludeAttributesDef:
                            OperationExcludeAttributes excludeAttributesOp = OperationExcludeAttributesPersistence.ToData(operation as CdmOperationExcludeAttributes, resOpt, options);
                            operations.Add(excludeAttributesOp);
                            break;
                        case CdmObjectType.OperationArrayExpansionDef:
                            OperationArrayExpansion arrayExpansionOp = OperationArrayExpansionPersistence.ToData(operation as CdmOperationArrayExpansion, resOpt, options);
                            operations.Add(arrayExpansionOp);
                            break;
                        case CdmObjectType.OperationCombineAttributesDef:
                            OperationCombineAttributes combineAttributesOp = OperationCombineAttributesPersistence.ToData(operation as CdmOperationCombineAttributes, resOpt, options);
                            operations.Add(combineAttributesOp);
                            break;
                        case CdmObjectType.OperationRenameAttributesDef:
                            OperationRenameAttributes renameAttributesOp = OperationRenameAttributesPersistence.ToData(operation as CdmOperationRenameAttributes, resOpt, options);
                            operations.Add(renameAttributesOp);
                            break;
                        case CdmObjectType.OperationReplaceAsForeignKeyDef:
                            OperationReplaceAsForeignKey replaceAsForeignKeyOp = OperationReplaceAsForeignKeyPersistence.ToData(operation as CdmOperationReplaceAsForeignKey, resOpt, options);
                            operations.Add(replaceAsForeignKeyOp);
                            break;
                        case CdmObjectType.OperationIncludeAttributesDef:
                            OperationIncludeAttributes includeAttributesOp = OperationIncludeAttributesPersistence.ToData(operation as CdmOperationIncludeAttributes, resOpt, options);
                            operations.Add(includeAttributesOp);
                            break;
                        case CdmObjectType.OperationAddAttributeGroupDef:
                            OperationAddAttributeGroup addAttributeGroupOp = OperationAddAttributeGroupPersistence.ToData(operation as CdmOperationAddAttributeGroup, resOpt, options);
                            operations.Add(addAttributeGroupOp);
                            break;
                        default:
                            OperationBase baseOp = new OperationBase();
                            baseOp.Type = OperationTypeConvertor.OperationTypeToString(CdmOperationType.Error);
                            operations.Add(baseOp);
                            break;
                    }
                }
            }

            return new Projection
            {
                Explanation = instance.Explanation,
                Source = source,
                Operations = operations,
                Condition = instance.Condition,
                RunSequentially = instance.RunSequentially
            };
        }
    }
}

File: objectModel/Python/cdm/persistence/cdmfolder/projections/projection_persistence.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See License.txt in the project root for license information.

from typing import TYPE_CHECKING, Union

from cdm.enums import CdmObjectType, CdmLogCode
from cdm.enums.cdm_operation_type import CdmOperationType, OperationTypeConvertor
from cdm.objectmodel import CdmOperationBase, CdmEntityReference
from cdm.persistence.cdmfolder.projections.operation_add_attribute_group_persistence import \
    OperationAddAttributeGroupPersistence
from cdm.persistence.cdmfolder.projections.operation_add_count_attribute_persistence import OperationAddCountAttributePersistence
from cdm.persistence.cdmfolder.projections.operation_add_supporting_persistence import OperationAddSupportingAttributePersistence
from cdm.persistence.cdmfolder.projections.operation_add_type_persistence import OperationAddTypeAttributePersistence
from cdm.persistence.cdmfolder.projections.operation_exclude_attributes_persistence import OperationExcludeAttributesPersistence
from cdm.persistence.cdmfolder.projections.operation_array_expansion_persistence import OperationArrayExpansionPersistence
from cdm.persistence.cdmfolder.projections.operation_combine_attributes_persistence import \
    OperationCombineAttributesPersistence
from cdm.persistence.cdmfolder.projections.operation_rename_attributes_persistence import OperationRenameAttributesPersistence
from cdm.persistence.cdmfolder.projections.operation_replace_as_foreign_key_persistence import \
    OperationReplaceAsForeignKeyPersistence
from cdm.persistence.cdmfolder.projections.operation_include_attributes_persistence import OperationIncludeAttributesPersistence

from cdm.persistence.cdmfolder.types import Projection, EntityReference
from cdm.persistence.cdmfolder.types.projections.operation_base import OperationBase
from cdm.utilities.logging import logger

if TYPE_CHECKING:
    from cdm.objectmodel import CdmCorpusContext, CdmProjection
    from cdm.utilities import ResolveOptions, CopyOptions

_TAG = 'ProjectionPersistence'


class ProjectionPersistence:
    """Projection persistence"""

    @staticmethod
    def from_data(ctx: 'CdmCorpusContext', data: Union[str, 'EntityReference', 'Projection']) -> 'CdmProjection':
        from cdm.persistence.cdmfolder.entity_reference_persistence import EntityReferencePersistence

        if not data:
            return None

        projection = ctx.corpus.make_object(CdmObjectType.PROJECTION_DEF)

        source = EntityReferencePersistence.from_data(ctx, data.get('source'))

        if data.explanation:
            projection.explanation = data.explanation

        if data.condition:
            projection.condition = data.condition

        projection.run_sequentially = data.runSequentially

        if data.operations is not None:
            operation_jsons = data.operations
            for operation_json in operation_jsons:
                type = operation_json.get('$type')

                if type == 'addCountAttribute':
                    add_count_attribute_op = OperationAddCountAttributePersistence.from_data(ctx, operation_json)
                    projection.operations.append(add_count_attribute_op)
                elif type == 'addSupportingAttribute':
                    add_supporting_attribute_op = OperationAddSupportingAttributePersistence.from_data(ctx, operation_json)
                    projection.operations.append(add_supporting_attribute_op)
                elif type == 'addTypeAttribute':
                    add_type_attribute_op = OperationAddTypeAttributePersistence.from_data(ctx, operation_json)
                    projection.operations.append(add_type_attribute_op)
                elif type == 'excludeAttributes':
                    exclude_attributes_op = OperationExcludeAttributesPersistence.from_data(ctx, operation_json)
                    projection.operations.append(exclude_attributes_op)
                elif type == 'arrayExpansion':
                    array_expansion_op = OperationArrayExpansionPersistence.from_data(ctx, operation_json)
                    projection.operations.append(array_expansion_op)
                elif type == 'combineAttributes':
                    combine_attributes_op = OperationCombineAttributesPersistence.from_data(ctx, operation_json)
                    projection.operations.append(combine_attributes_op)
                elif type == 'renameAttributes':
                    rename_attributes_op = OperationRenameAttributesPersistence.from_data(ctx, operation_json)
                    projection.operations.append(rename_attributes_op)
                elif type == 'replaceAsForeignKey':
                    replace_as_foreign_key_op = OperationReplaceAsForeignKeyPersistence.from_data(ctx, operation_json)
                    projection.operations.append(replace_as_foreign_key_op)
                elif type == 'includeAttributes':
                    include_attributes_op = OperationIncludeAttributesPersistence.from_data(ctx, operation_json)
                    projection.operations.append(include_attributes_op)
                elif type == 'addAttributeGroup':
                    add_attribute_group_op = OperationAddAttributeGroupPersistence.from_data(ctx, operation_json)
                    projection.operations.append(add_attribute_group_op)
                else:
                    logger.error(ctx, _TAG, ProjectionPersistence.from_data.__name__, None, CdmLogCode.ERR_PERSIST_PROJ_INVALID_OPS_TYPE, type)

        projection.source = source

        return projection

    @staticmethod
    def to_data(instance: 'CdmProjection', res_opt: 'ResolveOptions', options: 'CopyOptions') -> 'Projection':
        from cdm.persistence.cdmfolder.entity_reference_persistence import EntityReferencePersistence

        if not instance:
            return None

        if instance.source and isinstance(instance.source, str):
            source = instance.source
        elif instance.source and instance.source.named_reference and instance.source.explicit_reference == None:
            source = instance.source.named_reference
        elif instance.source and isinstance(instance.source, CdmEntityReference):
            source = EntityReferencePersistence.to_data(instance.source, res_opt, options)

        operations = None
        if instance.operations and len(instance.operations) > 0:
            operations = []
            for operation in instance.operations:
                if operation.object_type == CdmObjectType.OPERATION_ADD_COUNT_ATTRIBUTE_DEF:
                    add_count_attribute_op = OperationAddCountAttributePersistence.to_data(operation, res_opt, options)
                    operations.append(add_count_attribute_op)
                elif operation.object_type == CdmObjectType.OPERATION_ADD_SUPPORTING_ATTRIBUTE_DEF:
                    add_supporting_attribute_op = OperationAddSupportingAttributePersistence.to_data(operation, res_opt, options)
                    operations.append(add_supporting_attribute_op)
                elif operation.object_type == CdmObjectType.OPERATION_ADD_TYPE_ATTRIBUTE_DEF:
                    add_type_attribute_op = OperationAddTypeAttributePersistence.to_data(operation, res_opt, options)
                    operations.append(add_type_attribute_op)
                elif operation.object_type == CdmObjectType.OPERATION_EXCLUDE_ATTRIBUTES_DEF:
                    exclude_attributes_op = OperationExcludeAttributesPersistence.to_data(operation, res_opt, options)
                    operations.append(exclude_attributes_op)
                elif operation.object_type == CdmObjectType.OPERATION_ARRAY_EXPANSION_DEF:
                    array_expansion_op = OperationArrayExpansionPersistence.to_data(operation, res_opt, options)
                    operations.append(array_expansion_op)
                elif operation.object_type == CdmObjectType.OPERATION_COMBINE_ATTRIBUTES_DEF:
                    combine_attributes_op = OperationCombineAttributesPersistence.to_data(operation, res_opt, options)
                    operations.append(combine_attributes_op)
                elif operation.object_type == CdmObjectType.OPERATION_RENAME_ATTRIBUTES_DEF:
                    rename_attributes_op = OperationRenameAttributesPersistence.to_data(operation, res_opt, options)
                    operations.append(rename_attributes_op)
                elif operation.object_type == CdmObjectType.OPERATION_REPLACE_AS_FOREIGN_KEY_DEF:
                    replace_as_foreign_key_op = OperationReplaceAsForeignKeyPersistence.to_data(operation, res_opt, options)
                    operations.append(replace_as_foreign_key_op)
                elif operation.object_type == CdmObjectType.OPERATION_INCLUDE_ATTRIBUTES_DEF:
                    include_attributes_op = OperationIncludeAttributesPersistence.to_data(operation, res_opt, options)
                    operations.append(include_attributes_op)
                elif operation.object_type == CdmObjectType.OPERATION_ADD_ATTRIBUTE_GROUP_DEF:
                    add_attribute_group_op = OperationAddAttributeGroupPersistence.to_data(operation, res_opt, options)
                    operations.append(add_attribute_group_op)
                else:
                    base_op = OperationBase()
                    base_op.type = OperationTypeConvertor._operation_type_to_string(CdmOperationType.ERROR)
                    operations.append(base_op)

        obj = Projection()
        obj.explanation = instance.explanation
        obj.source = source
        obj.operations = operations
        obj.condition = instance.condition
        obj.runSequentially = instance.run_sequentially

        return obj
